File: SRChoco/SRHotkeyManager/SRHotKeyManager.m
//
//  SRHotKeyManager.m
//  SRChoco
//
//  Created by Heeseung Seo on 2014. 11. 10..
//  Copyright (c) 2014 Seorenn. All rights reserved.
//

#import "SRHotKeyManager.h"

#if TARGET_OS_IPHONE
#warning SRHotKeyManager was disabled because target platform is iOS.
#else

#import <Carbon/Carbon.h>

#define BOOLSTR(bv) ((bv) ? @"true":@"false")

#pragma mark - SRHotKey Class

@interface SRHotKey ()
@property (nonatomic, readonly) UInt32 modifiers;
@end

@implementation SRHotKey

- (id)initWithKeyCode:(UInt32)keyCode command:(BOOL)command control:(BOOL)control option:(BOOL)option shift:(BOOL)shift {
    self = [super init];
    if (self) {
        self.keyCode = keyCode;
        self.command = command;
        self.control = control;
        self.option = option;
        self.shift = shift;
    }
    return self;
}

- (UInt32)modifiers {
    UInt32 result = 0;
    
    if (self.command) result += cmdKey;
    if (self.control) result += controlKey;
    if (self.option) result += optionKey;
    if (self.shift) result += shiftKey;
    
    return result;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"<SRHotKey Command[%@] Control[%@] Option[%@] Shift[%@] Code[%d]>", BOOLSTR(self.command), BOOLSTR(self.control), BOOLSTR(self.option), BOOLSTR(self.shift), self.keyCode];
}

@end

#pragma mark - SRGlobalHotKeyManager Class

@interface SRGlobalHotKeyManager ()
@property (nonatomic, strong) SRHotKey *hotKey;
@property (nonatomic, strong) SRGlobalHotKeyHandler handler;
@end

OSStatus SRGlobalHotKeyManagerHandler(EventHandlerCallRef nextHandler, EventRef theEvent, void *userData) {
    SRGlobalHotKeyManager *manager = [SRGlobalHotKeyManager sharedManager];
    if (manager.handler) {
        manager.handler();
    }

    return noErr;
}

EventHotKeyRef g_hotKeyRef;

@implementation SRGlobalHotKeyManager

+ (SRGlobalHotKeyManager *)sharedManager {
    static SRGlobalHotKeyManager *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[SRGlobalHotKeyManager alloc] init];
    });
    return instance;
}

- (id)init {
    self = [super init];
    if (self) {
        EventTypeSpec eventType;
        eventType.eventClass = kEventClassKeyboard;
        eventType.eventKind = kEventHotKeyPressed;
        
        InstallApplicationEventHandler(&SRGlobalHotKeyManagerHandler, 1, &eventType, NULL, NULL);
    }
    return self;
}

- (void)registerWithHotKey:(SRHotKey *)hotKey handler:(SRGlobalHotKeyHandler)handler {
    self.hotKey = hotKey;
    self.handler = handler;
    
    EventHotKeyID hotKeyID;
    hotKeyID.signature = 'srgh';
    hotKeyID.id = 0;
    
    RegisterEventHotKey(hotKey.keyCode, hotKey.modifiers, hotKeyID, GetApplicationEventTarget(), 0, &g_hotKeyRef);
}

- (void)unregisterHotKey {
    UnregisterEventHotKey(g_hotKeyRef);
    self.hotKey = nil;
    self.handler = nil;
}

@end

#pragma mark - SRHotKeyManager Class

@implementation SRHotKeyManager
// TODO
@end

#endif

File: SRChoco/SRHotkeyManager/SRHotkeyManager.swift
//
// SRHotkeyManager.swift
// SRChoco
//
// Created by Seorenn.
// Copyright (c) 2014 Seorenn. All rights reserved.
//

#if os(OSX)
import AppKit
import Carbon
#else
import Foundation
#endif

typealias SRHotkeyHandler = (hotkey: SRHotkey) -> ()

class SRHotkey: NSObject, Printable {
    var command = false
    var control = false
    var option = false
    var shift = false
    var keycode: UInt32 = 0
    var handler: SRHotkeyHandler?
    
    init(keycode: UInt32, command: Bool, control: Bool, option: Bool, shift: Bool, handler: SRHotkeyHandler) {
        self.keycode = keycode
        self.command = command
        self.control = control
        self.option = option
        self.shift = shift
        self.handler = handler
        super.init()
    }
    
    override var description: String {
        var modifiers: [String] = []
        
        if self.command { modifiers.append("Command") }
        if self.control { modifiers.append("Control") }
        if self.option { modifiers.append("Option") }
        if self.shift { modifiers.append("Shift") }
        
        var modifiersString = ""
        if modifiers.count > 0 {
            modifiersString = modifiers.stringByJoining("+")
        }
        
        return "<SRHotkey \(modifiersString) [\(self.keycode)]"
    }
    
    var modifiers: UInt32 {
        var value: UInt32 = 0
        if self.command { value += cmdKey }
        if self.control { value += controlKey }
        if self.option { value += optionKey }
        if self.shift { value += shiftKey }
        
        return value
    }
}

class SRHotkeyManager: NSObject, Printable {
    private let signature = "srhm"
    
    // MARK: - Singleton Factory

    struct StaticInstance {
        static var dispatchToken: dispatch_once_t = 0
        static var instance: SRHotkeyManager?
    }
    
    class func sharedManager() -> SRHotkeyManager {
        dispatch_once(&StaticInstance.dispatchToken) {
            StaticInstance.instance = SRHotkeyManager()
        }
        return StaticInstance.instance!
    }
    
    // MARK: - Initializers
    override init() {
        // TODO
        super.init()
    }
    
    override var description: String {
        // TODO
        return "SRHotkeyManager"
    }

    // MARK: - Public Methods

}